src/httpattack/httpattack.go

```diff
@@ -0,0 +1,273 @@
+		ReadTimeout     *time.Duration `mapstructure:"read_timeout"`
+		WriteTimeout    *time.Duration `mapstructure:"write_timeout"`
+		IdleTimeout     *time.Duration `mapstructure:"idle_timeout"`
+		MaxIdleConns    *int           `mapstructure:"max_idle_connections"`
+		ProxyURLs       string         `mapstructure:"proxy_urls"`
+	}
+
+	if err := utils.Decode(clientCfg, &clientConfig); err != nil && debug {
+		log.Printf("Failed to parse job client, ignoring: %v", err)
+	}
+
+	timeout := 90 * time.Second
+	if clientConfig.Timeout != nil {
+		timeout = *clientConfig.Timeout
+	}
+
+	readTimeout := timeout
+	if clientConfig.ReadTimeout != nil {
+		readTimeout = *clientConfig.ReadTimeout
+	}
+
+	writeTimeout := timeout
+	if clientConfig.WriteTimeout != nil {
+		writeTimeout = *clientConfig.WriteTimeout
+	}
+
+	idleTimeout := timeout
+	if clientConfig.IdleTimeout != nil {
+		idleTimeout = *clientConfig.IdleTimeout
+	}
+
+	maxIdleConns := 1000
+	if clientConfig.MaxIdleConns != nil {
+		maxIdleConns = *clientConfig.MaxIdleConns
+	}
+
+	tlsConfig := &tls.Config{
+		InsecureSkipVerify: true,
+	}
+	if clientConfig.TLSClientConfig != nil {
+		tlsConfig = clientConfig.TLSClientConfig
+	}
+
+	proxy := func() string { return "" }
+	proxylist := templates.ParseAndExecute(clientConfig.ProxyURLs, nil)
+	if proxylist != "" {
+		if debug {
+			log.Printf("List of proxies: %s", proxylist)
+		}
+
+		var proxyURLs = strings.Split(proxylist, ",")
+
+		if debug {
+			log.Printf("proxyURLs: %v", proxyURLs)
+		}
+
+		// Return random proxy from the list
+		proxy = func() string {
+			if len(proxyURLs) == 0 {
+				return ""
+			}
+
+			return proxyURLs[rand.Intn(len(proxyURLs))]
+		}
+	}
+
+	defaultProxy := fasthttpproxy.FasthttpProxyHTTPDialerTimeout(timeout)
+	if proxyURL != "" {
+		defaultProxy = fasthttpproxy.FasthttpSocksDialer(proxyURL)
+	}
+
+	cli := &fasthttp.Client{
+		ReadTimeout:                   readTimeout,
+		WriteTimeout:                  writeTimeout,
+		MaxConnDuration:               timeout,
+		MaxIdleConnDuration:           idleTimeout,
+		MaxConnsPerHost:               maxIdleConns,
+		NoDefaultUserAgentHeader:      true, // Don't send: User-Agent: fasthttp
+		DisableHeaderNamesNormalizing: true, // If you set the case on your headers correctly you can enable this
+		DisablePathNormalizing:        true,
+		TLSConfig:                     tlsConfig,
+		// increase DNS cache time to an hour instead of default minute
+		Dial: newFastHTTPProxyDial(proxy, timeout, defaultProxy),
+	}
+
+	return &HttpAttackClient{client: cli}
+}
+
+type HttpAttackClient struct {
+	client *fasthttp.Client
+}
+
+// Attacking provided path. Return transfered data
+func (c *HttpAttackClient) Attack(ctx context.Context,
+	method, path, body string,
+	headers map[string]string, cookies map[string]string, debug bool) (int, error) {
+
+	methodTpl, pathTpl, bodyTpl, headerTpls, cookieTpls, err := prepareHTTPRequestTemplates(method, path, body, headers, cookies)
```

Comment:
the point of splitting parsing and execution of templates was performance - parsing is quite expensive and shouldn't be done in a loop. Here you essentially break that idea completely by moving parsing back into the loop iteration